### PR TITLE
Flow builder refactor getContents, handoff and action 

### PR DESCRIPTION
**Pull request by Iru89**
## Description
Refactor the getContents function to be able to get the contents in several ways (through the node, Id or code), refactor the handoff logic creating the FlowHandoff class. Also refactor the action to apply the changes in the handoff and make it more readable. 

## Context
Have several ways to get the contents through an id or the code. This way we have clearer functions to obtain the contents through an Id in the actions that are inside the bot and through the code for the campaign plugin.

## Approach taken / Explain the design
Refactor several things. 
The function getContents is now called getContentsByNode and only accepts one node, you can no longer pass an id. The return is no longer an object, it now returns the array of contents (the handoffNode disappears and will be treated as a normal content).
I have created the functions getContentsById and getContentsByCode that receive an Id or a code and end up calling getContentsByNode in this way we can obtain the contents in 3 ways with the Id, the code or the node. For the campaigns plugin we will use the getContentsByCode.

Refactor all handoff logic.
I have created the FlowHandoff class just like we have for FlowText, FlowCarousel etc. This class has a function fromHubtypeCMS that creates an instance of FlowHandoff from the handoff node. It also has the logic to get the queue, the onFinishPayload and do the handoff.

Refactor the logic into action.
I have created functions with clearer names to indicate which node is obtained in each case getNodeByUserInput getFallbackNode.
When the handoffNode disappears I apply a filter before passing the contents to the render to do the handoff and eliminate that node. 

## Testing

The pull request...

- has unit tests
- has integration tests
- doesn't need tests because... **[provide a description]**

### CODE REVIEW

File: packages/botonic-plugin-flow-builder/src/index.ts [hunk 1]
```diff
@@ -61,7 +61,14 @@ export default class BotonicPluginFlowBuilder implements Plugin {
     })
   }
 
-  async post(_request: PluginPostRequest): Promise<void> {}
```
Comment (AinaVendrell) [inline packages/botonic-plugin-flow-builder/src/index.ts:64]: isn't it mandatory to define the post?

File: packages/botonic-plugin-flow-builder/package.json [hunk 2]
```diff
@@ -29,6 +29,7 @@
     "README.md"
   ],
   "devDependencies": {
+    "@types/axios": "^0.14.0",
```
Comment (vanbasten17) [inline packages/botonic-plugin-flow-builder/package.json:32]: `axios` is on its 1.X.X version, should we update it?

**Review by vanbasten17 — APPROVED** [review]

File: packages/botonic-plugin-flow-builder/package.json (hunk 2 above)
Comment (Iru89) [reply packages/botonic-plugin-flow-builder/package.json:32]: You are right, I don't know why but the other day I got an error that I didn't have the arios types. But now I have checked it and in the version we are using it is not necessary to import the @types separately.

Comment (Iru89) [reply packages/botonic-plugin-flow-builder/package.json:32]: I will remove this dependency

File: packages/botonic-plugin-flow-builder/src/index.ts (hunk 1 above)
Comment (Iru89) [reply packages/botonic-plugin-flow-builder/src/index.ts:64]: no in the latest version of botonic core the pre and post functions are optional, but you have to implement at least one in order to implement Plugin.

File: packages/botonic-plugin-flow-builder/src/action.tsx [hunk 3]
```diff
@@ -75,3 +70,30 @@ export class FlowBuilderMultichannelAction extends FlowBuilderAction {
...
+): HtNodeWithContent | undefined {
+  if (request.session.is_first_interaction) {
+    const startNode = cmsApi.getStartNode()
+    return startNode
+  }
+
+  if (request.input.data) {
+    const intentNode = cmsApi.getNodeByIntent(request.input, locale)
+    const keywordNode = cmsApi.getNodeByKeyword(request.input.data, locale)
+    return intentNode ?? keywordNode
```
Comment (AinaVendrell) [inline packages/botonic-plugin-flow-builder/src/action.tsx:87]: if `intentNode` is defined we could return it before computing the `keywordNode`

File: packages/botonic-plugin-flow-builder/src/content-fields/flow-handoff.tsx [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  | HtImageNode
+  | HtVideoNode
+  | HtCarouselNode
+  | HtWhatsappButtonListNode
+
+export class FlowHandoff extends ContentFieldsBase {
+  public code: string
+  public queue?: HtQueueLocale
+  public onFinishPayload?: string
+  public hanndoffAutoAssign: boolean
```
Comment (AinaVendrell) [inline packages/botonic-plugin-flow-builder/src/content-fields/flow-handoff.tsx:29]: `hanndoffAutoAssign` --> `handoffAutoAssign`

File: packages/botonic-plugin-flow-builder/src/content-fields/flow-handoff.tsx [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    if (!payloadId) return undefined
+
+    const actionPayload = cmsApi.getNodeById(payloadId)
+
+    return (actionPayload as HtPayloadNode).content.payload
+  }
+
+  async doHandoff(request: ActionRequest) {
+    // @ts-ignore
+    const handOffBuilder = new HandOffBuilder(request.session)
```
Comment (AinaVendrell) [inline packages/botonic-plugin-flow-builder/src/content-fields/flow-handoff.tsx:74]: why this `@ts-ignore`

File: packages/botonic-plugin-flow-builder/src/action.tsx [hunk 6]
```diff
@@ -1,50 +1,39 @@
...
+      ) as HtNodeWithContent
     }
 
     if (!targetNode) {
-      targetNode = flowBuilderPlugin.cmsApi.getFallbackNode(
-        alternateFallbackMessage
-      )
-      alternateFallbackMessage = !alternateFallbackMessage
+      targetNode = getFallbackNode(flowBuilderPlugin.cmsApi, request)
     }
```
Comment (AinaVendrell) [inline packages/botonic-plugin-flow-builder/src/action.tsx:34]: extra: maybe we could move this to a function like: 
```
function getTargetNode(
  cmsApi: FlowBuilderApi,
  locale: string,
  request: ActionRequest
) {
  const contentId = request.input.payload
  let targetNode: HtNodeWithContent | undefined
  if (!contentId) {
    targetNode = getNodeByUserInput(cmsApi, locale, request)
  } else {
    targetNode = cmsApi.getNodeById(contentId) as HtNodeWithContent
  }
  if (targetNode) {
    return targetNode
  }
  return getFallbackNode(cmsApi, request)
}
```

**Review by AinaVendrell — CHANGES_REQUESTED** [review]

File: packages/botonic-plugin-flow-builder/src/action.tsx [hunk 7]
```diff
@@ -75,3 +70,30 @@ export class FlowBuilderMultichannelAction extends FlowBuilderAction {
...
 }
+
+function getNodeByUserInput(
+  cmsApi: FlowBuilderApi,
+  locale: string,
+  request: ActionRequest
+): HtNodeWithContent | undefined {
+  if (request.session.is_first_interaction) {
+    const startNode = cmsApi.getStartNode()
+    return startNode
```
Comment (vanbasten17) [inline packages/botonic-plugin-flow-builder/src/action.tsx:81]:
```suggestion
return cmsApi.getStartNode()
```

File: packages/botonic-plugin-flow-builder/src/action.tsx (hunk 3 above)
Comment (vanbasten17) [reply packages/botonic-plugin-flow-builder/src/action.tsx:87]: yep, this is a symptom that we are facing an issue with typing somewhere

File: packages/botonic-plugin-flow-builder/src/action.tsx [hunk 8]
```diff
@@ -75,3 +70,30 @@ export class FlowBuilderMultichannelAction extends FlowBuilderAction {
...
+  }
+
+  if (request.input.data) {
+    const intentNode = cmsApi.getNodeByIntent(request.input, locale)
+    const keywordNode = cmsApi.getNodeByKeyword(request.input.data, locale)
+    return intentNode ?? keywordNode
+  }
+
+  return undefined
```
Comment (vanbasten17) [inline packages/botonic-plugin-flow-builder/src/action.tsx:90]: if !(none of the above)
return undefined
return intentNode ?? keyWordNode

File: packages/botonic-plugin-flow-builder/src/action.tsx [hunk 9]
```diff
@@ -75,3 +70,30 @@ export class FlowBuilderMultichannelAction extends FlowBuilderAction {
...
+}
+
+function getFallbackNode(cmsApi: FlowBuilderApi, request: ActionRequest) {
+  const isFirstFallbackOption =
+    request.session.user.extra_data.isFirstFallbackOption || true
+  const fallbackNode = cmsApi.getFallbackNode(isFirstFallbackOption)
+  request.session.user.extra_data.isFirstFallbackOption = !isFirstFallbackOption
+  return fallbackNode
```
Comment (vanbasten17) [inline packages/botonic-plugin-flow-builder/src/action.tsx:98]: nit: do use early return?

File: packages/botonic-plugin-flow-builder/src/content-fields/content-fields-base.ts [hunk 10]
```diff
@@ -21,4 +22,12 @@ export abstract class ContentFieldsBase {
     const result = video.find(v => v.locale === locale)
     return result?.url ?? ''
   }
+
+  static getQueueByLocale(
+    locale: string,
+    queues: HtQueueLocale[]
+  ): HtQueueLocale | undefined {
+    const queue = queues.find(queue => queue.locale === locale)
+    return queue
```
Comment (vanbasten17) [inline packages/botonic-plugin-flow-builder/src/content-fields/content-fields-base.ts:31]:
```suggestion
return queues.find(queue => queue.locale === locale)
```

File: packages/botonic-plugin-flow-builder/src/content-fields/flow-handoff.tsx (hunk 5 above)
Comment (vanbasten17) [reply packages/botonic-plugin-flow-builder/src/content-fields/flow-handoff.tsx:74]: idk, we have to improve overall types

**Review by vanbasten17 — APPROVED** [review]
LGTM!!

File: packages/botonic-plugin-flow-builder/src/content-fields/flow-handoff.tsx (hunk 5 above)
Comment (Iru89) [reply packages/botonic-plugin-flow-builder/src/content-fields/flow-handoff.tsx:74]: I have to update the Session type

File: packages/botonic-plugin-flow-builder/src/action.tsx [hunk 11]
```diff
@@ -54,9 +43,15 @@ export class FlowBuilderAction extends React.Component<FlowBuilderActionProps> {
       await flowBuilderPlugin.trackEvent(request, contents[0].code)
     }
 
-    if (handoffNode) await doHandoff(request, locale, handoffNode)
+    const renderContents = contents.filter(async content => {
+      if (content instanceof FlowHandoff) {
```
[the remark this answers is not in the document]
Comment (Iru89) [reply packages/botonic-plugin-flow-builder/src/action.tsx:47]: I don't quite understand which early return you are referring to. Here I'm using a filter to remove the contents of handoff so I return a false after executing the doHandoff function

File: packages/botonic-plugin-flow-builder/src/action.tsx [hunk 12]
```diff
@@ -1,50 +1,39 @@
...
-      const intentNode = flowBuilderPlugin.cmsApi.getNodeByIntent(
-        request.input,
-        locale
-      )
-      const keywordNode = flowBuilderPlugin.cmsApi.getNodeByKeyword(
-        request.input.data,
-        locale
-      )
-      targetNode = intentNode ?? keywordNode ?? targetNode
+    if (!contentId) {
```
[the remark this answers is not in the document]
Comment (Iru89) [reply packages/botonic-plugin-flow-builder/src/action.tsx:24]: I have applied the refactor proposed by Aina

**Comment by vanbasten17** [comment]
@Iru89 go ahead with the ones which already have 1 approve

**Review by AinaVendrell — APPROVED** [review]

**Pull request merged**